src/Builder.php

```diff
@@ -0,0 +1,170 @@
+<?php
+
+namespace LaraCrafts\UrlShortener;
+
+use GuzzleHttp\Psr7\Uri;
+use InvalidArgumentException;
+use LaraCrafts\UrlShortener\Concerns\CustomDomains;
+use LaraCrafts\UrlShortener\Concerns\CustomSuffixes;
+use LaraCrafts\UrlShortener\Concerns\ToString;
+use LaraCrafts\UrlShortener\Concerns\ToUri;
+use LaraCrafts\UrlShortener\Contracts\Client;
+use LaraCrafts\UrlShortener\Contracts\UnsupportedOperationException;
+use Psr\Http\Message\UriInterface;
+
+class Builder
+{
+    protected $client;
+    protected $options;
+    protected $uri;
+
+    /**
+     * Create a new shortening builder instance.
+     *
+     * @param \LaraCrafts\UrlShortener\Contracts\Client $client
+     * @param \Psr\Http\Message\UriInterface|string $uri
+     * @param array $options
+     * @return void
+     */
+    public function __construct(Client $client, $uri, array $options)
```

Comment:
As it is a builder I would not put options as a parameter, at least not as a required one.